auditlog/event_builder.go

```diff
@@ -0,0 +1,203 @@
+}
+
+type Filter struct {
+	RequestHeaderWhiteList  []string `json:"request_header"`
+	RequestBodyWhiteList    []string `json:"request_body"`
+	ResponseHeaderWhiteList []string `json:"response_header"`
+	ResponseBodyWhiteList   []string `json:"response_body"`
+}
+
+// UnmarshalJSON implements the Unmarshaler interface for the EventBuilder struct.
+func (b *EventBuilder) UnmarshalJSON(raw []byte) error {
+	var err error
+
+	// An additional struct like in rule.go.
+	var bb struct {
+		URLPattern          string `json:"url_pattern"`
+		Method              string `json:"http_method"`
+		Filter              Filter `json:"filter"`
+		DescriptionTemplate string `json:"description_template"`
+		r                   *regexp.Regexp
+	}
+
+	if err = json.Unmarshal(raw, &bb); err != nil {
+		return err
+	}
+
+	*b = bb
+	b.r, err = regexp.Compile(b.URLPattern)
+
+	return err
+}
+
+// Match method verifies if this builder must be applied to request with given url and method.
+func (b *EventBuilder) Match(url, method string) bool {
+	if b.Method != method {
+		return false
+	}
+
+	if !b.r.MatchString(url) {
+		return false
+	}
+
+	return true
+}
+
+// Build method performs filtering of data using rules from config.
+func (b *EventBuilder) Build(req *http.Request, resp *http.Response, err error) (*Event, error) {
+	e := NewEvent()
+
+	if req != nil {
+		e.Meta["method"] = req.Method
+		e.Meta["url"] = req.URL.String()
+		e.Meta["user_ip"] = req.RemoteAddr
+
+		if sess, ok := req.Context().Value(proxy.ContextKeySession).(*authn.AuthenticationSession); ok {
+			e.Meta["user_id"] = sess.Subject
+		}
+
+		e.RequestHeader = filterHeader(req.Header, b.Filter.RequestHeaderWhiteList)
+		e.RequestBody = filterBody(req.Body, b.Filter.RequestBodyWhiteList)
+	}
+
+	if err != nil {
+		e.OathkeeperError = err
+	}
+
+	if resp != nil {
+		e.Meta["status_code"] = strconv.Itoa(resp.StatusCode)
+
+		e.ResponseHeader = filterHeader(resp.Header, b.Filter.ResponseHeaderWhiteList)
+		e.ResponseBody = filterBody(resp.Body, b.Filter.ResponseBodyWhiteList)
+	}
+
+	// TODO generate Description.
+	e.Description = ""
+
+	return &e, nil
+}
+
+func filterBody(b io.ReadCloser, wl []string) map[string]interface{} {
+	result := make(map[string]interface{})
+
+	if b == nil {
+		return result
+	}
+
+	bb, err := ioutil.ReadAll(b)
+	if err != nil {
+		return result
+	}
+
+	var body map[string]interface{}
+	if err = json.Unmarshal(bb, &body); err != nil {
+		return result
+	}
+
+NextWhitelistItem:
+	for _, key := range wl {
+		var value interface{} = body
+		for _, k := range strings.Split(key, ".") {
+			var ok bool
+
+			m, ok := value.(map[string]interface{})
+			if !ok {
+				continue NextWhitelistItem
+			}
+
+			value, ok = m[k]
+			if !ok {
+				continue NextWhitelistItem
+			}
+		}
+		result[key] = value
+	}
+
+	return result
+}
+
+func filterHeader(h http.Header, wl []string) map[string]string {
+	result := make(map[string]string)
+	for _, key := range wl {
+		result[key] = h.Get(key)
```

Comment:
So, the better behavior is not to insert a key if no value met? I think I agree with you.

Meh, I forgot that it is allowed to have several key-value pairs with the same key in JSON. Now I think it would be better to use a third-party  JSON parser here.

Reply:
➕

Reply:
Done.